### PR TITLE
Add option to create geometric bodies (spheres, prisms, ellipsoids, etc) and compute their effect.

### PR DESCRIPTION
**Pull request by joa-quim**
Tested for a couple of them (bodies) but not all.

### CODE REVIEW

**Review by PaulWessel — COMMENTED** [review]
You cannot break the **+**_modifier_ syntax since we will auto-compute these from long-forms, eventually.  Things like **-M+bell** is not a modifier since **bell** is a word.  It has to be, e.g., **-M+s**_bell_ where **+s** means shape.  We can then later handle **-Mshape**=_bell_ in long-form and auto-convert to **-M+s**_bell_.  So need to be **+s**cylinder (see typo), **+s**cone, **+s**prism, etc.

Type Alaternative.

**Comment by joa-quim** [comment]
OK, **-M+ssphere** it is.

**Review by PaulWessel — COMMENTED** [review]
You need to merge master into this first, no.  Otherwise, I see you are reverting several things in master related to recent bug fixes, e.g., in gmt_io.c, grdcut.c, grdedit.c.
Your test script is missing the **+**s modifier.
The listing of bell,height, etc should be _italics_ not **bold** since these are parameters, and you have several references to <hight> instead of *height* in that section (i.e., there are lots of these).  Pirimid -> Pyramid.

**Comment by joa-quim** [comment]
I had *merge master* but had conflicts and ended up no committing all master changes in this branch. Had also forgotten to commit the cube_mag.sh test update to +s.

File: doc/rst/source/supplements/potential/gmtgravmag3d.rst [hunk 1]
```diff
@@ -12,7 +12,7 @@ Synopsis
 
 .. include:: ../../common_SYN_OPTs.rst_
 
-**gmt gravmag3d** *xyz_file* |-T|\ **v**\ *vert_file* OR |-T|\ **r\|s**\ *raw_file*
+**gmt gravmag3d** *xyz_file* |-T|\ **v**\ *vert_file* OR |-T|\ **r\|s**\ *raw_file* OR |-M|\ **+**\ *body,params*
```
Comment (PaulWessel) [inline doc/rst/source/supplements/potential/gmtgravmag3d.rst:15]:
```suggestion
**gmt gravmag3d** *xyz_file* |-T|\ **v**\ *vert_file* OR |-T|\ **r\|s**\ *raw_file* OR |-M|\ **+s**\ *body,params*
```

File: doc/rst/source/supplements/potential/gmtgravmag3d.rst [hunk 2]
```diff
@@ -60,6 +60,30 @@ Required Arguments
 **-G**\ *outgrid*
     Output the gravity or magnetic anomaly at nodes of this grid file.
 
+.. _-M:
+
+**-M+s**\ *body,params* (Alaternative to **-Tr**\ /**-Ts**). Create geometric bodies and compute their grav/mag effect.
```
Comment (PaulWessel) [inline doc/rst/source/supplements/potential/gmtgravmag3d.rst:65]:
```suggestion
**-M+s**\ *body,params* (An alternative to **-Tr**\ /**-Ts**). Create geometric bodies and compute their grav/mag effect.
```

File: doc/rst/source/supplements/potential/gmtgravmag3d.rst [hunk 3]
```diff
@@ -60,6 +60,30 @@ Required Arguments
...
+    Select among one or more of the following bodies, where *x0* & *y0* represent the horizontal coordinates
+    of the body center [default to 0,0 positive up], *npts* is the number of points that a circle is discretized
+    and *n_slices* apply when bodies are made by a pile of slices. For example Spheres and Ellipsoids are made of
+    *2 x n_slices* and Bells have *n_slices* [Default 5]. It is even possible to select more than one body. For example
+    **-M+s**\ *prism,1/1/1/-5/-10/1*\ **+s**\ *sphere,1/-5* computes the effect of a prism and a sphere. Unfortunately there is
+    no current way of selecting distinct densities or magnetic parameters for each body.
+
+      - *bell,height/sx/sy/z0[/x0/y0/n_sig/npts/n_slices]* Gaussian of height *height* with caracteristic STDs *sx* and *sy*. The base width (at depth *z0*) is controled by the number of sigmas (*n_sig*) [Default = 2]\n");
```
Comment (PaulWessel) [inline doc/rst/source/supplements/potential/gmtgravmag3d.rst:73]:
```suggestion
      - *bell,height/sx/sy/z0[/x0/y0/n_sig/npts/n_slices]* Gaussian of height *height* with characteristic STDs *sx* and *sy*. The base width (at depth *z0*) is controled by the number of sigmas (*n_sig*) [Default = 2]\n");
```

File: doc/rst/source/supplements/potential/gmtgravmag3d.rst [hunk 4]
```diff
@@ -130,11 +154,11 @@ specified unit to meter.  If your grid is geographic, convert distances to meter
 Examples
 --------
 
-Suppose you ...
-
-   ::
+To compute the magnetic anomaly of a cube of unity side located at 5 meters depth and centered at -10,1 in a domain
```
Comment (PaulWessel) [inline doc/rst/source/supplements/potential/gmtgravmag3d.rst:157]:
```suggestion
To compute the magnetic anomaly of a cube of unit sides located at 5 meters depth and centered at -10,1 in a domain
```

File: src/potential/solids.c [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	Ctrl->raw_mesh[m].t3[0] =  ellipse[j3][i3].x;
+	Ctrl->raw_mesh[m].t3[1] = -ellipse[j3][i3].y;
+	Ctrl->raw_mesh[m].t3[2] = -z3;
+}
+
+int prism(struct GMT_CTRL *GMT, struct GMTGRAVMAG3D_CTRL *Ctrl, int nb) {
+	double a, b, c, z_c, x0, y0, z_top, z_bot;
+	int i_tri = Ctrl->n_raw_triang;
+
+	/* Note, position [9] bellow holds current body number */
```
Comment (PaulWessel) [inline src/potential/solids.c:30]:
```suggestion
	/* Note, position [9] below holds current body number */
```

File: src/potential/solids.c [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+	Ctrl->raw_mesh[i_tri].t2[0] =  a/2 + x0;	Ctrl->raw_mesh[i_tri].t2[1] =  b/2 - y0;	Ctrl->raw_mesh[i_tri].t2[2] = z_bot;
+	Ctrl->raw_mesh[i_tri].t3[0] =  a/2 + x0;	Ctrl->raw_mesh[i_tri].t3[1] = -b/2 - y0;	Ctrl->raw_mesh[i_tri].t3[2] = z_bot;
+
+	Ctrl->n_raw_triang += 12;
+	return 0;
+}
+
+int five_psoid(struct GMT_CTRL *GMT, struct GMTGRAVMAG3D_CTRL *Ctrl, int body_type, int nb, bool cone, bool piram, bool sino, bool hemi) {
+/*	Constructs either a sphere, ellipsoid, cone, pyramid, or a bell
+/*	as a union of triagular facets. Returns number of triangles. */
```
Comment (PaulWessel) [inline src/potential/solids.c:109]:
```suggestion
/*	as a union of triangular facets. Returns number of triangles. */
```

**Comment by joa-quim** [comment]
Ping

**Comment by PaulWessel** [comment]
Think you have to do all those Mimi commits

**Comment by PaulWessel** [comment]
Mini

**Comment by seisman** [comment]
@joa-quim manually fixed the typos in the commit 70c119613070f17d546493c7bb66ab1f642af214.

You don't have to do that. You can just click the "Commit suggestions" button.

**Comment by seisman** [comment]
Ping @joa-quim  
```
Following bash scripts may not have execute permission:
1: ./test/potential/cube_mag.sh
```

**Comment by joa-quim** [comment]
They are committed.

**Comment by PaulWessel** [comment]
OK, I guess you did not use the buttons we set up for you to click commit as they all show up there, so harder to see if they were fixed.  Also, per @seisman the script does not have executable permission so please fix that so we can approve.
Sorry, slow day with family insisting on morning swim

**Comment by joa-quim** [comment]
The script has the ``+x`` permission on my side. But the commit doesn't believe it.

**Comment by PaulWessel** [comment]
OK, will fix if needed.

**Review by PaulWessel — APPROVED** [review]

**Comment by seisman** [comment]
> The script has the `+x` permission on my side. But the commit doesn't believe it.

FYI, I added the `+x` permission to the script and pushed the change directly to the master branch.

**Comment by seisman** [comment]
@joa-quim This PR introduces some compiler warnings:
```
../../src/potential/gmtgravmag3d.c:243:7: warning: unused variable 'ptr' [-Wunused-variable]
        char ptr[GMT_LEN256] = {""};
             ^
In file included from ../../src/potential/gmtgravmag3d.c:667:
../../src/potential/solids.c:37:3: warning: '/*' within block comment [-Wcomment]
                /* first triangle */
                ^
../../src/potential/solids.c:108:1: warning: '/*' within block comment [-Wcomment]
/*      as a union of triangular facets. Returns number of triangles. */
^
```

**Pull request merged**